### ISSUE TITLE
`cachedImplicit` fails for `LabelledGeneric` when used inside companion object

### ISSUE THREAD

**Issue by notxcain** [milessabin/shapeless#390]
Here is the code
```
import shapeless._

sealed trait Root
case class A(s: String) extends Root
case class B(i: Int) extends Root

object Root {
  implicit val lgen: LabelledGeneric[Root] = cachedImplicit //fails
  implicit val gen: Generic[Root] = cachedImplicit //compiles
}
```

**Comment by milessabin** [comment]
I think that part of the problem here is that the type annotations are actually incorrect: in both cases they're explicitly discarding the refinement, ie. `type Repr = ...` which is essential for these values to be usable.

There's no straightforward way of adding the refinement to the annotation, so I think that for now at least the unannotated, hence inferred, form will have to do,

```scala
object Root {
  implicit val gen = the[Generic[Root]]
  implicit val lgen = the[LabelledGeneric[Root]]
}
```

**Comment by milessabin** [comment]
I'm going leave this open for further investigation, but given that there is a workaround I'll put this until the next release.

**Comment by notxcain** [comment]
Ok, thanks for the workaround!

**Comment by milessabin** [comment]
Closing this in favour of #508.

**Comment by milessabin** [comment]
I have a fairly ugly workaround for the type annotations part of this issue,

```scala
import shapeless._, test._

case class Foo(i: Int, s: String)

object Foo {
  import Generic._ // This is the workaround from #508
  val g1 = { // nb. type of g1 must be inferred
    def fooGen = ??? // Here we shadow the following defn so it isn't used in the Generic below
    Generic[Foo]
  }

  // Now we can define both the value and the type of fooGen in terms of the
  // non-implicit g1 ...
  implicit val fooGen: Generic.Aux[Foo, g1.Repr] = g1 
}

object Demo extends App {
  val gen = Generic[Foo]
  val foo = Foo(23, "foo")
  val fooRepr = gen.to(foo)
  typed[Int :: String :: HNil](fooRepr)
  assert(fooRepr.head+13 == 36)
}
```

**Comment by notxcain** [comment]
I see what you did here :) And yes it's pretty ugly. I like where it goes in #527

**Comment by milessabin** [comment]
Reopening because it's about to be fixed properly :-)

**Comment by milessabin** [comment]
As of 49bb32ff5051b9b2650b37106d9925b7b92fcb43 the following should work as expected,

```scala
import shapeless._

sealed trait Root
case class A(s: String) extends Root
case class B(i: Int) extends Root

object Root {
  val lgen0 = cachedImplicit[LabelledGeneric[Root]]
  implicit val lgen: LabelledGeneric.Aux[Root, lgen0.Repr] = lgen0

  val gen0 = cachedImplicit[Generic[Root]]
  implicit val gen: Generic.Aux[Root, gen0.Repr] = gen0
}
```

**Comment by clhodapp** [comment]
Whoa, careful there, @milessabin. If you do it that way, `lgen0` will find `lgen` and take `null` (similar for `gen0` and `gen`). I think you mean something more like:
```scala
import shapeless._

sealed trait Root
case class A(s: String) extends Root
case class B(i: Int) extends Root

object Root {
  implicit val lgen = cachedImplicit[LabelledGeneric[Root]]

  implicit val gen = cachedImplicit[Generic[Root]]
}
```

**Comment by clhodapp** [comment]
Also, as I noted (somewhere): `lgen eq gen` here, since `LabelledGeneric[Root] <:< Generic[Root]` and the second `cachedImplicit` finds `lgen`.

**Comment by clhodapp** [comment]
That said, the issue is fixed now, as `cachedImplicit` is now sort of like `the` now in that it preserves refinements.

**Comment by milessabin** [comment]
@notxcain no, I meant it exactly as written. See the test [here](https://github.com/milessabin/shapeless/blob/master/core/src/test/scala/shapeless/implicits.scala#L88-L99).

**Comment by clhodapp** [comment]
Hrm. So I tried some code in a repl  both with the RC1 code and the code from my fork that had the compat hack and and the results weren't what either of us probably expected: https://gist.github.com/clhodapp/497bd765e7c8015558b9.

**Comment by milessabin** [comment]
Yes, that result genuinely surprised me. It turns out the not having `LabelledGeneric[T]` extend `Generic[T]` fixes the problem. I think that's probably the right thing to do, but want to canvas other opinions on this first.

**Comment by milessabin** [comment]
Fixed in e8926d163adc7015403c8296b22ed71827c585a0.

**Comment by milessabin** [comment]
Some new solutions to the cached `Generic` type ascription problem based on `the` and `TypeOf` described from [here](https://github.com/milessabin/shapeless/pull/692#issuecomment-283956735) on.
